### ISSUE TITLE
shared_credentials_file for aws provider does not expand `~/<path>`

### ISSUE THREAD

**Issue by tmulhern3** [hashicorp/terraform#6193]
The `aws` provider does not expand the user's home directory `~/` for the `shared_credentials_file` argument. See below:

`provider "aws" {
    shared_credentials_file  = "~/.aws/credentials"
    profile                  = "Profile"
    region = "${var.region}"
}`

Error message: `No valid credential sources found for AWS Provider.`

Please note that and error does not occur when supplying the full path instead of ~/.

If this issue appears to affect multiple resources, it may be an issue with Terraform's core, so please mention this.

This appears to have happened in the past as well: - GH-3856

**Comment by fgimian** [comment]
Would really love to see this too as it would make it easier for multiple team members to run Terraform without having to modify this path to suit their home directory.  For now, perhaps using the AWS_SHARED_CREDENTIALS_FILE environment variable is the best way to go.

**Comment by Techbrunch** [comment]
This issue looks like a duplicate of #11268, it also looks like it was fixed in #11277.

According to the [documentation](https://www.terraform.io/docs/configuration/interpolation.html):

```
pathexpand(string) - Returns a filepath string with ~ expanded to the home directory. Note: This will create a plan diff between two different hosts, unless the filepaths are the same.
```

So it should be possible to do this:

```hcl
provider "aws" {
  region                   = "us-west-2"
  shared_credentials_file  = "${pathexpand(~/.aws/creds)}"
  profile                  = "customprofile"
}
```
